knack/deprecation.py

```diff
@@ -0,0 +1,179 @@
+        return self._color + self._message + colorama.Fore.RESET
+
+
+# pylint: disable=too-many-instance-attributes
+class Deprecated(object):
+
+    @staticmethod
+    def ensure_new_style_deprecation(cli_ctx, kwargs, object_type):
+        """ Helper method to make the previous string-based deprecate_info kwarg
+            work with the new style. """
+        deprecate_info = kwargs.get('deprecate_info', None)
+        if isinstance(deprecate_info, Deprecated):
+            deprecate_info.object_type = object_type
+        elif isinstance(deprecate_info, str):
+            deprecate_info = Deprecated(cli_ctx, redirect=deprecate_info, object_type=object_type)
+        kwargs['deprecate_info'] = deprecate_info
+        return deprecate_info
+
+    def __init__(self, cli_ctx=None, object_type='', target=None, redirect=None, hide=False, expiration=None,
+                 tag_func=None, message_func=None):
+        """ Create a collection of deprecation metadata.
+
+        :param cli_ctx: The CLI context associated with the deprecated item.
+        :type cli_ctx: knack.cli.CLI
+        :param object_type: A label describing the type of object being deprecated.
+        :type: object_type: str
+        :param target: The name of the object being deprecated.
+        :type target: str
+        :param redirect: The alternative to redirect users to in lieu of the deprecated item. If omitted it, there is
+                         no alternative.
+        :type redirect: str
+        :param hide: A boolean or CLI version at or above-which the deprecated item will no longer appear
+                     in help text, but will continue to work. Warnings will be displayed if the deprecated
+                     item is used.
+        :type hide: bool OR str
+        :param expiration: The CLI version at or above-which the deprecated item will no longer work.
+        :type expiration: str
+        :param tag_func: Callable which returns the desired unformatted tag string for the deprecated item.
+                         Omit to use the default.
+        :type tag_func: callable
+        :param message_func: Callable which returns the desired unformatted message string for the deprecated item.
+                             Omit to use the default.
+        :type message_func: callable
+        """
+        self.cli_ctx = cli_ctx
+        self.object_type = object_type
+        self.target = target
+        self.redirect = redirect
+        self.hide = hide
+        self.expiration = expiration
+
+        def _default_get_message(self):
+            message = "This {} has been deprecated and will be removed ".format(self.object_type)
+            if self.expiration:
+                message += "in version '{}'. ".format(self.expiration)
+            else:
+                message += 'in a future release. '
+            message += "Use '{}' instead. ".format(self.redirect) if self.redirect else ''
+            return message.lstrip()
```

Comment:
A majority of this is replicated elsewhere

Reply:
Good point. I'll see if I can abstract this out.

Reply:
Reviewing these cases, I'm not sure there's much value in templatizing this part.